File: manga_backend/app/schemas/manga.py
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class TagBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    TagId: uuid.UUID
    GroupName: str | None = None
    NameEn: str | None = None


class AltTitle(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    LangCode: str | None = None
    AltTitle: str | None = None


class Description(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    LangCode: str | None = None
    Description: str | None = None


class LinkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    Provider: str | None = None
    Url: str | None = None


class StatisticsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    Follows: int | None = 0
    AverageRating: float | None = 0
    BayesianRating: float | None = 0


class CreatorOut(BaseModel):
    id: uuid.UUID | None = None
    name: str | None = None
    role: str | None = None  # author / artist


class MangaListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    MangaId: uuid.UUID
    TitleEn: str | None = None
    Status: str | None = None
    Year: int | None = None
    ContentRating: str | None = None
    PublicationDemographic: str | None = None
    cover_url: str | None = None
    stats: StatisticsOut | None = None


class MangaDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    MangaId: uuid.UUID
    Type: str | None = None
    TitleEn: str | None = None
    Status: str | None = None
    Year: int | None = None
    ContentRating: str | None = None
    PublicationDemographic: str | None = None
    OriginalLanguage: str | None = None
    LastChapter: str | None = None
    LastVolume: str | None = None
    CreatedAt: datetime | None = None
    UpdatedAt: datetime | None = None
    cover_url: str | None = None
    stats: StatisticsOut | None = None
    tags: list[TagBrief] = []
    alt_titles: list[AltTitle] = []
    descriptions: list[Description] = []
    links: list[LinkOut] = []
    creators: list[CreatorOut] = []
    available_languages: list[str] = []


class RelatedMangaOut(BaseModel):
    RelatedId: uuid.UUID
    relation_type: str | None = None
    related_label: str | None = None
    title: str | None = None
    cover_url: str | None = None

File: manga_frontend/src/types/manga.ts
import type { UUID } from "./common";

export interface Statistics {
  Follows?: number | null;
  AverageRating?: number | null;
  BayesianRating?: number | null;
}

export interface TagBrief {
  TagId: UUID;
  GroupName?: string | null;
  NameEn?: string | null;
}

export interface AltTitle {
  LangCode?: string | null;
  AltTitle?: string | null;
}

export interface Description {
  LangCode?: string | null;
  Description?: string | null;
}

export interface LinkOut {
  Provider?: string | null;
  Url?: string | null;
}

export interface CreatorOut {
  id?: UUID;
  name?: string | null;
  role?: string | null;
}

export interface MangaListItem {
  MangaId: UUID;
  TitleEn?: string | null;
  Status?: string | null;
  Year?: number | null;
  ContentRating?: string | null;
  PublicationDemographic?: string | null;
  cover_url?: string | null;
  stats?: Statistics | null;
}

export interface MangaDetail extends MangaListItem {
  Type?: string | null;
  OriginalLanguage?: string | null;
  LastChapter?: string | null;
  LastVolume?: string | null;
  CreatedAt?: string | null;
  UpdatedAt?: string | null;
  tags: TagBrief[];
  alt_titles: AltTitle[];
  descriptions: Description[];
  links: LinkOut[];
  creators: CreatorOut[];
  available_languages: string[];
}

export interface RelatedManga {
  RelatedId: UUID;
  relation_type?: string | null;
  related_label?: string | null;
  title?: string | null;
  cover_url?: string | null;
}

export interface MangaListParams {
  page?: number;
  limit?: number;
  sort?: string;
  status?: string;
  content_rating?: string;
  demographic?: string;
  year?: number;
}

export interface AdvancedSearchParams extends MangaListParams {
  q?: string;
  include_tags?: string;
  exclude_tags?: string;
  year_from?: number;
  year_to?: number;
  original_lang?: string;
}

export interface TagGroup {
  group_name: string;
  tags: TagBrief[];
}

export interface Cover {
  cover_id?: UUID;
  manga_id?: UUID;
  volume?: string | null;
  locale?: string | null;
  fileName?: string | null;
  cover_url?: string | null;
}
